gh-120321: Make gi_frame_state transitions atomic in FT build

colesbury opened the pull request:
This makes generator frame state transitions atomic in the free threading build, which avoids segfaults when trying to execute a generator from multiple threads concurrently.

There are still a few operations that aren't thread-safe and may crash if performed concurrently on the same generator/coroutine:

 * Accessing gi_yieldfrom/cr_await/ag_await
 * Accessing gi_frame/cr_frame/ag_frame
 * Async generator operations


<!-- gh-issue-number: gh-120321 -->
* Issue: gh-120321
<!-- /gh-issue-number -->

colesbury committed:
gh-120321: Make gi_frame_state transitions atomic in FT build

This makes generator frame state transitions atomic in the free
threading build, which avoids segfaults when trying to execute
a generator from multiple threads concurrently.

There are still a few operations that aren't thread-safe and may crash
if performed concurrently on the same generator/coroutine:

 * Accessing gi_yieldfrom/cr_await/ag_await
 * Accessing gi_frame/cr_frame/ag_frame
 * Async generator operations

colesbury committed:
Add some tests

colesbury committed:
Merge branch 'main' into gh-120321-gen-atomic

colesbury committed:
Remove unused import

mpage commented on Python/bytecodes.c:
```diff
@@ -1425,6 +1422,10 @@ dummy_func(
+            FT_ATOMIC_STORE_INT8_RELEASE(gen->gi_frame_state, FRAME_SUSPENDED + oparg);
+#ifdef Py_GIL_DISABLED
+            ((_PyThreadStateImpl *)tstate)->gen_last_frame_state = FRAME_SUSPENDED + oparg;
+#endif
```
Would it make sense to factor this out into a helper function? It's duplicated in `ceval.c`. Also, having helper to update both would reduce the chance that we update one without the other.

colesbury replied:
I've refactored it out. I also moved the functions to `ceval_macros.h` from `pycore_genobject.h` because:

1) they are only used in ceval.c and bytecodes.c
2) I want to avoid adding additional includes to `pycore_genobject.h` because I think it's pulled in by internal headers that some third-party projects tend to include.

mpage commented on Include/internal/pycore_genobject.h:
```diff
@@ -21,7 +34,30 @@ PyGenObject *_PyGen_GetGeneratorFromFrame(_PyInterpreterFrame *frame)
+    }
+#endif
+    if (gen->gi_frame_state < FRAME_EXECUTING) {
+        gen->gi_frame_state = FRAME_EXECUTING;
```
Do these loads need to use atomics (in the FT build)?

colesbury replied:
No, this is the case where the generator is uniquely referenced so we don't need to use atomics. I added a comment.

kumaraditya303 commented on Objects/genobject.c:
```diff
@@ -254,18 +234,29 @@ gen_send_ex2(PyGenObject *gen, PyObject *arg, PyObject **presult,
 
-    gen->gi_frame_state = FRAME_EXECUTING;
+#if defined(Py_GIL_DISABLED) && defined(Py_DEBUG)
+    ((_PyThreadStateImpl *)tstate)->gen_last_frame_state = FRAME_EXECUTING;
```
why is this set only in debug builds?

colesbury replied:
I changed this to `#if defined(Py_GIL_DISABLED)`. It's not really needed at all, but it'll help catch bugs in case some ceval code forgets to set `gen_last_frame_state` before returning.

In other words, when `_PyEval_EvalFrame` returns, `gen_last_frame_state` must be either `FRAME_SUSPENDED[_YIELD_FROM]` or `FRAME_CLEARED` and this helps differentiate this call to `_PyEval_EvalFrame` from earlier calls in the same thread.

kumaraditya303 commented on Objects/genobject.c:
```diff
@@ -282,46 +273,101 @@ gen_send_ex2(PyGenObject *gen, PyObject *arg, PyObject **presult,
-    return gen_send_ex((PyGenObject*)gen, arg, 0, 0);
+    PyObject *result;
+    PyGenObject *gen = _PyGen_CAST(op);
+    if (gen_send_ex(_PyGen_CAST(op), arg, &result) == PYGEN_RETURN) {
```
```suggestion
    if (gen_send_ex(gen, arg, &result) == PYGEN_RETURN) {
```

kumaraditya303 commented on Objects/genobject.c:
```diff
@@ -442,15 +495,80 @@ gen_close(PyObject *self, PyObject *args)
+
+    if (PyExceptionClass_Check(typ))
+        PyErr_NormalizeException(&typ, &val, &tb);
+
```
```suggestion
    if (PyExceptionClass_Check(typ)) {
        PyErr_NormalizeException(&typ, &val, &tb);
    }
```

kumaraditya303 commented on Include/internal/pycore_genobject.h:
```diff
@@ -21,7 +34,30 @@ PyGenObject *_PyGen_GetGeneratorFromFrame(_PyInterpreterFrame *frame)
+// Mark the generator as executing. Returns true if the state was changed,
+// false if it was already executing or finished.
+static inline bool
+_PyGen_SetExecuting(PyGenObject *gen)
```
since this function can fail, maybe name it as `_PyGen_TrySetExecuting` like other free-threading functions which can fail like `TryIncref`

colesbury committed:
Update Objects/genobject.c

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

colesbury committed:
Update Objects/genobject.c

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

colesbury committed:
Changes from review

colesbury committed:
Merge branch 'main' into gh-120321-gen-atomic

colesbury committed:
Remove extra whitespace

kumaraditya303 commented on Objects/genobject.c:
```diff
@@ -145,10 +153,6 @@ _PyGen_Finalize(PyObject *self)
 {
-    if (gen->gi_frame_state == FRAME_CLEARED)
-        return;
-
-    gen->gi_frame_state = FRAME_CLEARED;
```
```c
assert(gen->gi_frame_state == FRAME_CLEARED);
```
to ensure that caller has set the frame state correctly

kumaraditya303 commented on Objects/genobject.c:
```diff
@@ -36,6 +36,14 @@ static PyObject* async_gen_athrow_new(PyAsyncGenObject *, PyObject *);
+# define _Py_GEN_SET_FRAME_STATE(gen, expected, state) \
+    _Py_atomic_compare_exchange_int8(&(gen)->gi_frame_state, &expected, (state))
+#else
+# define _Py_GEN_SET_FRAME_STATE(gen, expected, state) \
```
```suggestion
# define _Py_GEN_TRY_SET_FRAME_STATE(gen, expected, state) \
```

markshannon commented on Objects/genobject.c:
```diff
@@ -254,18 +234,29 @@ gen_send_ex2(PyGenObject *gen, PyObject *arg, PyObject **presult,
+    // Grab the last frame state from the thread state instead of the
+    // generator, as it may have changed if another thread resumed this
+    // generator.
+    int8_t frame_state = ((_PyThreadStateImpl *)tstate)->gen_last_frame_state;
```
How can another thread resume this generator? This thread is executing it, so no other thread can run it.

colesbury replied:
This thread is no longer executing the generator once it sets the frame state to suspended in `YIELD_VALUE`. Between that line in `bytecodes.c` and here, another thread can immediately start executing the same generator. That isn't true with the GIL because there are no opportunities for reentrancy or GIL release between `YIELD_VALUE` and the return from `_PyEval_Frame`. That lack of opportunities for reentrancy is also why it's safe to store it on the thread state.

markshannon replied:
OK, I see what this issue is.
What you want is an additional return value, to distinguish between a `yield` and a `return`.

I think the code would be clearer, if `gen_last_frame_state` field were renamed to something like `generator_return_kind`. Set it to `RETURN` before the call, and set it to `YIELD` in `YIELD_VALUE`.
Instead of testing for `FRAME_STATE_SUSPENDED(frame_state)` test for `generator_return_kind != RETURN`

Also add a comment about *why* it is needed.

colesbury replied:
I've renamed the field and adjusted the condition.

However, we need to set `gen_last_frame_state` in `gen_clear_frame` (or elsewhere in `RETURN_VALUE`). It's not enough to set it to `RETURN` before the `_PyEval_EvalFrame` call because generators can send to other generator. The "inner" generator will overwrite the `gen_last_frame_state`. (See the test failures on https://github.com/python/cpython/pull/142599/commits/4b82b8cbe566dae85aaf93ff885e8c861c207475)

markshannon commented on Objects/genobject.c:
```diff
@@ -282,46 +273,101 @@ gen_send_ex2(PyGenObject *gen, PyObject *arg, PyObject **presult,
+           FRAME_STATE_SUSPENDED(frame_state));
+
+    if (!_Py_GEN_SET_FRAME_STATE(gen, frame_state, FRAME_EXECUTING)) {
+        goto retry;
```
No need for a goto. You can use `do ... while (!_Py_GEN_SET_FRAME_STATE(gen, frame_state, FRAME_EXECUTING));`

colesbury replied:
I've switched from gotos to do/while loops

markshannon reviewed (commented):
The changes to bytecodes.c and ceval/cevalmacros look good.

I'm puzzled by the `gen_last_frame_state` field on the thread state.
It seems unnecessary and, if it is necessary, broken in case of reentrancy.

colesbury committed:
Merge branch 'main' into gh-120321-gen-atomic

colesbury committed:
Rename _Py_GEN_SET_FRAME_STATE to _Py_GEN_TRY_SET_FRAME_STATE

colesbury committed:
Switch from goto to do/while

colesbury committed:
Merge branch 'main' into gh-120321-gen-atomic

colesbury committed:
Rename field to generator_return_kind

Also use it in both the default and FT builds.

colesbury committed:
Set generator_return_kind in clear_gen_frame

Generator calls may be nested so it's not sufficient to set the field
before the call to PyEval_Frame.

colesbury committed:
Fix stray typo

mpage reviewed (approved):
LGTM!

kumaraditya303 commented on Objects/genobject.c:
```diff
@@ -281,47 +275,97 @@ gen_send_ex2(PyGenObject *gen, PyObject *arg, PyObject **presult,
+            else if (arg) {
+                /* `gen` is an exhausted generator:
+                only return value if called from send(). */
+                *presult = Py_NewRef(Py_None);
```
```suggestion
                *presult = Py_None;
```
`None` is immortal

colesbury committed:
Update Objects/genobject.c

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

colesbury committed:
Merge branch 'main' into gh-120321-gen-atomic

cdce8p commented on Objects/genobject.c:
```diff
@@ -370,43 +414,43 @@ is_resume(_Py_CODEUNIT *instr)
+    int8_t frame_state = FT_ATOMIC_LOAD_INT8_RELAXED(gen->gi_frame_state);
+    do {
+        if (frame_state == FRAME_CREATED) {
+            if (!_Py_GEN_TRY_SET_FRAME_STATE(gen, frame_state, FRAME_CLEARED)) {
+                continue;
+            }
```
This check seems to trigger a new compiler warning during build.
```
Objects/genobject.c:426:17: warning: code will never be executed [-Wunreachable-code]
                continue;
```

Tested on MacOS with gcc.

colesbury replied:
A few questions:

* Are you sure it's actually gcc? Which version?
* Are you adding extra compiler flags? I don't think we compile with `-Wunreachable-code` and I don't think it's enabled by `-Wall` and `-Wextra`.

colesbury replied:
Ah, I figured out why I wasn't seeing it.

* We enable `-Wunreachable-code` on Clang only and choose to do so as part of the `./configure`. It's only enabled in non-debug builds
* I think `./configure -C` caches the value, so I didn't see it when switching from debug to non-debug build.

cdce8p replied:
Yes, it was clang. I should have checked it more carefully. Sorry about that.

colesbury replied:
https://github.com/python/cpython/pull/143022

colesbury replied:
Thanks for the catching this
